### DISCUSSION TITLE
AsyncStorage update order guarantee?

### DISCUSSION THREAD

**Discussion by jaskp** [pmndrs/jotai#1073] [Q&A]
I'd like to use Jotai in my RN app and I need persistence. I found an example in the docs, but this part got me a bit concerned

```ts
(get, set, update) => {
  const nextValue = typeof update === 'function' ? update(get(baseAtom)) : update
  set(baseAtom, nextValue)
  AsyncStorage.setItem(key, JSON.stringify(nextValue))
}
```

The promise basically gets thrown away. Is there any guarantee at all that (rapid) consecutive updates of the atom will result in the correct order of updates to storage? 

At first I thought Zustand handled this diferently, because the example in the docs returns the promise instead of throwing it away:
```ts
async (name: string, value: string): Promise<void> => {
  console.log(name, "with value", value, "has been saved");
  await set(name, value)
}
```
But upon checking the source, it looks like it's thrown away anyway.
```ts
api.setState = (state, replace) => {
  savedSetState(state, replace)
  void setItem()
}
```

Any thoughts on this? Maybe Redux's flushing is safer? I honestly don't know.

Thanks!

**Comment by dai-shi** [comment]
jotai atom's `write` function is invoked in sync. There's no trick behind it. You can assume it's just a function call.

> The promise basically gets thrown away.

Is your concern not having `await` for `set(baseAtom, nextValue)`? If `baseAtom`'s write function returns a promise, then having `await` is reasonable and better.

**Reply by jaskp** [reply]
Not exactly my concern. I mean if I set the derived atom to `x` and right afterwards set it to `y`, what's guaranteeing that AsyncStorage will resolve `setItem(key, x)` before `setItem(key, y)` ? I was expecting some mechanisms in Jotai that queues these setters and runs them sequentially, because now they would be parallel promises, wouldn't they?

**Reply by dai-shi** [reply]
No, unless you use promises in your atoms, all write functions should be sync. There are no such queues. What will be queued is rendering the result in React.
